recipes_source/inference_tuning_on_aws_graviton.rst

```diff
@@ -0,0 +1,360 @@
+    print(f"Using {device} device")
+
+
+3. Given linear layers are at the heart of several neural networks, including transformers, we take a linear layer for this demo. We define our neural network by subclassing ``nn.Module``, and initializing the layers in ``__init__``. We construct the network with a typical large language model parameters to match the real world scenario:
+
+.. code-block:: python
+
+  class MyNeuralNetwork(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.flatten = nn.Flatten()
+        self.linear_relu_stack = nn.Sequential(
+            nn.Linear(4096, 4096),
+            nn.ReLU(),
+            nn.Linear(4096, 11008),
+            nn.ReLU(),
+            nn.Linear(11008, 10),
+        )
+
+    def forward(self, x):
+        x = self.flatten(x)
+        logits = self.linear_relu_stack(x)
+        return logits
+
+4. Let's create an instance of ``MyNeuralNetwork``, and move it to the device:
+
+.. code-block:: python
+
+    model = MyNeuralNetwork().to(device)
+    print(model)
+
+Next, let's get the prediction probabilities by passing them through an instance of the ``nn.Softmax`` module:
+
+.. code-block:: python
+
+    X = torch.rand(1, 64, 64, device=device)
+    logits = model(X)
+    pred_probab = nn.Softmax(dim=1)(logits)
+    y_pred = pred_probab.argmax(1)
+    print(f"Predicted class: {y_pred}")
+
+output:
+
+.. code-block::
+
+    Predicted class: tensor([2])
+
+Our network functionality is verified. Next, we will profile the performance. Lets' check two different scenarios: small and large batch dimensions.
+
+**Scenario 1:** A larger batch dimension, for example 256:
+
+.. code-block:: python
+
+    # warm it up first and loop over multiple times to have enough execution time
+
+    X = torch.rand(256, 64, 64, device=device)
+
+    with torch.set_grad_enabled(False):
+        for _ in range(50):
+            model(X) #Warmup
+        with profile(activities=[ProfilerActivity.CPU]) as prof:
+            with record_function("mymodel_inference"):
+                for _ in range(100):
+                    model(X)
+
+    print(prof.key_averages().table(sort_by="self_cpu_time_total"))
+
+
+Following is the profiler output with the default PyTorch configuration:
+
+.. table:: profiler output with the torch default configuration
+   
+======================  ============   ===========  =============  ===========  ============  ============  
+                  Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
+======================  ============   ===========  =============  ===========  ============  ============
+           aten::addmm        97.61%       15.813s        98.61%       15.977s      53.255ms           300  
+       aten::clamp_min         1.09%     177.032ms         1.09%     177.032ms     885.160us           200  
+           aten::copy_         1.00%     162.054ms         1.00%     162.054ms     540.180us           300  
```

Comment:
` aten::copy_ ` for some reason creates a link in the HTML: 
<img width="884" alt="Screenshot 2023-12-26 at 7 56 08 PM" src="https://github.com/pytorch/tutorials/assets/5317992/bf433919-1b1a-4e8a-aeee-90f56f0a4ba7">
I wonder if the correct indentation will help with that.